store-gateway: caching large lists of label values causes panics

#### Describe the bug

Caching a list of more than 655360 label values causes panics. The root cause is golang issue 59172 which is a panic in encoding/gob.


<details>
<summary>panic</summary>

```
panic: runtime error: index out of range [655360] with length 655360 [recovered]
    panic: runtime error: index out of range [655360] with length 655360

goroutine 129499 [running]:
encoding/gob.catchError(0xc0468804f0)
    $GOROOT/src/encoding/gob/error.go:38 +0x6d
panic({0x2cc1640, 0xc03e175920})
    $GOROOT/src/runtime/panic.go:884 +0x213
encoding/gob.decStringSlice(0xc039c57020, {0x2894580?, 0xc039c499c0?, 0xa00000?}, 0xb1c85, {0x97?, 0x7f301a933a98?})
    $GOROOT/src/encoding/gob/dec_helpers.go:374 +0x1ef
encoding/gob.(*Decoder).decodeArrayHelper(0xc039c499c0?, 0xc039c57020, {0x2894580?, 0xc039c499c0?, 0x0?}, 0x3145840, 0xb1c85, {0x35156a0?, 0xc039c53be0}, 0x3145838)
    $GOROOT/src/encoding/gob/decode.go:540 +0xa3
encoding/gob.(*Decoder).decodeSlice(0x0?, 0x0?, {0x2894580?, 0xc039c499c0?, 0x0?}, 0x0?, {0x35156a0, 0xc039c53be0}, 0x0?)
    $GOROOT/src/encoding/gob/decode.go:669 +0x29f
encoding/gob.(*Decoder).decOpFor.func3(0x2bad0c0?, 0xc039c499c0?, {0x2894580?, 0xc039c499c0?, 0x0?})
    $GOROOT/src/encoding/gob/decode.go:887 +0x3a
encoding/gob.(*Decoder).decodeStruct(0x29bdb40?, 0xc039c570a0, {0x2bad0c0?, 0xc039c499c0?, 0x35154e0?})
    $GOROOT/src/encoding/gob/decode.go:494 +0x18b
encoding/gob.(*Decoder).decodeValue(0xc046880480, 0x451400?, {0x2832fa0?, 0xc039c499c0?, 0xc03a0b8301?})
    $GOROOT/src/encoding/gob/decode.go:1249 +0x24e
encoding/gob.(*Decoder).DecodeValue(0xc046880480, {0x2832fa0?, 0xc039c499c0?, 0x0?})
    $GOROOT/src/encoding/gob/decoder.go:229 +0x1b7
encoding/gob.(*Decoder).Decode(0xc046880480, {0x2832fa0?, 0xc039c499c0?})
    $GOROOT/src/encoding/gob/decoder.go:204 +0x165
github.com/grafana/mimir/pkg/storegateway.decodeSnappyGob({0xc046fc8000?, 0x25affcf94c473620?, 0xc001136c60?}, {0x2832fa0, 0xc039c499c0})
    //vendor/github.com/grafana/mimir/pkg/storegateway/snappy_gob_codec.go:38 +0x13c
github.com/grafana/mimir/pkg/storegateway.fetchCachedLabelValues({0x35398c8, 0xc03a08d770}, {0x355f800, 0xc001136c60}, {0xc0012353c7, 0x6}, {0x1, 0x88, 0x18, 0x27, ...}, ...)
    //vendor/github.com/grafana/mimir/pkg/storegateway/bucket.go:1369 +0x146
github.com/grafana/mimir/pkg/storegateway.blockLabelValues({0x35398c8, 0xc03a08d770}, 0xc001c8ec60, {0x352b018?, 0xc001235510}, 0xc0337374d1?, {0xc039c187e8, 0x3}, {0x4b8dc78, 0x0, ...}, ...)
    //vendor/github.com/grafana/mimir/pkg/storegateway/bucket.go:1248 +0x1cd
github.com/grafana/mimir/pkg/storegateway.(*BucketStore).LabelValues.func1()
    //vendor/github.com/grafana/mimir/pkg/storegateway/bucket.go:1195 +0x105
golang.org/x/sync/errgroup.(*Group).Go.func1()
    //vendor/golang.org/x/sync/errgroup/errgroup.go:75 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
    //vendor/golang.org/x/sync/errgroup/errgroup.go:72 +0xa5
```


</details>

#### To Reproduce

This test reproduces the bug 6da2b4c6ff989ec546f4d45a53b4fa4488bb45d2

#### Possible fixes

* wait for go 1.21 to be released with the fix
* avoid caching responses containing more than 100 000 label values
	* these are expensive and hopefully rare; so we can just not cache them and recompute next time around
* protobuf encoding
	* using a naive field of `repeated string values = 1` sometimes causes worse performance due to protobuf not preallocating slice size; this means we will likely want to introduce some modifications similar to how #3751 introduced `PreallocatingMetric`
	* <details>
		<summary>gob encoding vs autogenerated protobuf</summary>
		
		```
		goos: darwin
		goarch: arm64
		pkg: github.com/grafana/mimir/pkg/storegateway
		                                                                                                                           │  before.txt   │ use-protobuf-for-labevalues-b5d5fe24f.txt │
		                                                                                                                           │    sec/op     │       sec/op        vs base               │
		BucketStoreLabelValues/no_cache/10-series-matched-with-10-label-values-10                                                     679.6µ ± ∞ ¹         726.8µ ± ∞ ¹   +6.94% (p=0.029 n=4)
		BucketStoreLabelValues/no_cache/1000-series-matched-with-1000-label-values-10                                                 13.87m ± ∞ ¹         15.97m ± ∞ ¹  +15.19% (p=0.029 n=4)
		BucketStoreLabelValues/no_cache/1_000_000-series-matched-with-10-label-values-10                                              29.00m ± ∞ ¹         29.13m ± ∞ ¹        ~ (p=0.486 n=4)
		BucketStoreLabelValues/no_cache/1_000_000-series-matched-with-1-label-values-10                                               219.0µ ± ∞ ¹         228.6µ ± ∞ ¹        ~ (p=0.486 n=4)
		BucketStoreLabelValues/no_cache/1_000_000-series-matched-with-1_000_000-label-values-10                                       117.8m ± ∞ ¹         109.5m ± ∞ ¹        ~ (p=0.057 n=4)
		BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/10-series-matched-with-10-label-values-10                  115.9µ ± ∞ ¹         115.3µ ± ∞ ¹        ~ (p=0.886 n=4)
		BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1000-series-matched-with-1000-label-values-10              1.203m ± ∞ ¹         1.288m ± ∞ ¹        ~ (p=0.486 n=4)
		BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1_000_000-series-matched-with-10-label-values-10           15.09m ± ∞ ¹         15.12m ± ∞ ¹        ~ (p=0.886 n=4)
		BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1_000_000-series-matched-with-1-label-values-10            94.50µ ± ∞ ¹         87.21µ ± ∞ ¹   -7.71% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1_000_000-series-matched-with-1_000_000-label-values-10    114.4m ± ∞ ¹         106.5m ± ∞ ¹   -6.92% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/10-series-matched-with-10-label-values-10                                               73.76µ ± ∞ ¹         18.51µ ± ∞ ¹  -74.90% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1000-series-matched-with-1000-label-values-10                                          119.91µ ± ∞ ¹         87.98µ ± ∞ ¹  -26.63% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-10-label-values-10                                        73.94µ ± ∞ ¹         26.28µ ± ∞ ¹  -64.46% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1-label-values-10                                         72.29µ ± ∞ ¹         26.42µ ± ∞ ¹  -63.46% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1_000_000-label-values-10                                                      69.47m ± ∞ ¹
		geomean                                                                                                                       1.156m               1.192m        -22.88%
		¹ need >= 6 samples for confidence interval at level 0.95
		
		                                                                                                                           │   before.txt   │ use-protobuf-for-labevalues-b5d5fe24f.txt │
		                                                                                                                           │      B/op      │        B/op         vs base               │
		BucketStoreLabelValues/no_cache/10-series-matched-with-10-label-values-10                                                     804.2Ki ± ∞ ¹        794.5Ki ± ∞ ¹   -1.21% (p=0.029 n=4)
		BucketStoreLabelValues/no_cache/1000-series-matched-with-1000-label-values-10                                                 9.899Mi ± ∞ ¹        9.842Mi ± ∞ ¹   -0.57% (p=0.029 n=4)
		BucketStoreLabelValues/no_cache/1_000_000-series-matched-with-10-label-values-10                                              186.3Mi ± ∞ ¹        186.3Mi ± ∞ ¹   -0.01% (p=0.029 n=4)
		BucketStoreLabelValues/no_cache/1_000_000-series-matched-with-1-label-values-10                                              110.40Ki ± ∞ ¹        99.59Ki ± ∞ ¹   -9.79% (p=0.029 n=4)
		BucketStoreLabelValues/no_cache/1_000_000-series-matched-with-1_000_000-label-values-10                                       356.0Mi ± ∞ ¹        242.9Mi ± ∞ ¹  -31.77% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/10-series-matched-with-10-label-values-10                  52.69Ki ± ∞ ¹        42.62Ki ± ∞ ¹  -19.12% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1000-series-matched-with-1000-label-values-10              1.544Mi ± ∞ ¹        1.510Mi ± ∞ ¹   -2.16% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1_000_000-series-matched-with-10-label-values-10           148.1Mi ± ∞ ¹        148.1Mi ± ∞ ¹   -0.01% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1_000_000-series-matched-with-1-label-values-10            35.07Ki ± ∞ ¹        24.59Ki ± ∞ ¹  -29.90% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1_000_000-series-matched-with-1_000_000-label-values-10    356.0Mi ± ∞ ¹        242.9Mi ± ∞ ¹  -31.77% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/10-series-matched-with-10-label-values-10                                              51.391Ki ± ∞ ¹        8.461Ki ± ∞ ¹  -83.54% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1000-series-matched-with-1000-label-values-10                                           213.7Ki ± ∞ ¹        253.6Ki ± ∞ ¹  +18.63% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-10-label-values-10                                       50.951Ki ± ∞ ¹        8.006Ki ± ∞ ¹  -84.29% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1-label-values-10                                        51.043Ki ± ∞ ¹        6.969Ki ± ∞ ¹  -86.35% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1_000_000-label-values-10                                                      331.1Mi ± ∞ ¹
		geomean                                                                                                                       1.464Mi              1.322Mi        -39.13%
		¹ need >= 6 samples for confidence interval at level 0.95
		
		                                                                                                                           │  before.txt  │ use-protobuf-for-labevalues-b5d5fe24f.txt │
		                                                                                                                           │  allocs/op   │     allocs/op       vs base               │
		BucketStoreLabelValues/no_cache/10-series-matched-with-10-label-values-10                                                    1.910k ± ∞ ¹         1.766k ± ∞ ¹   -7.54% (p=0.029 n=4)
		BucketStoreLabelValues/no_cache/1000-series-matched-with-1000-label-values-10                                                28.25k ± ∞ ¹         28.07k ± ∞ ¹   -0.61% (p=0.029 n=4)
		BucketStoreLabelValues/no_cache/1_000_000-series-matched-with-10-label-values-10                                             1.726k ± ∞ ¹         1.581k ± ∞ ¹   -8.37% (p=0.029 n=4)
		BucketStoreLabelValues/no_cache/1_000_000-series-matched-with-1-label-values-10                                               967.0 ± ∞ ¹          820.0 ± ∞ ¹  -15.20% (p=0.029 n=4)
		BucketStoreLabelValues/no_cache/1_000_000-series-matched-with-1_000_000-label-values-10                                      3.000M ± ∞ ¹         3.000M ± ∞ ¹   -0.01% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/10-series-matched-with-10-label-values-10                  800.0 ± ∞ ¹          656.0 ± ∞ ¹  -18.00% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1000-series-matched-with-1000-label-values-10             12.65k ± ∞ ¹         12.49k ± ∞ ¹   -1.30% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1_000_000-series-matched-with-10-label-values-10          1.257k ± ∞ ¹         1.113k ± ∞ ¹  -11.46% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1_000_000-series-matched-with-1-label-values-10            646.0 ± ∞ ¹          499.0 ± ∞ ¹  -22.76% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1_000_000-series-matched-with-1_000_000-label-values-10   3.000M ± ∞ ¹         3.000M ± ∞ ¹   -0.01% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/10-series-matched-with-10-label-values-10                                              1343.0 ± ∞ ¹          185.0 ± ∞ ¹  -86.22% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1000-series-matched-with-1000-label-values-10                                          4.314k ± ∞ ¹         3.183k ± ∞ ¹  -26.22% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-10-label-values-10                                       1342.0 ± ∞ ¹          184.0 ± ∞ ¹  -86.29% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1-label-values-10                                        1330.0 ± ∞ ¹          166.0 ± ∞ ¹  -87.52% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1_000_000-label-values-10                                                     3.000M ± ∞ ¹
		geomean                                                                                                                      5.945k               5.540k        -40.56%
		¹ need >= 6 samples for confidence interval at level 0.95

		```
		
		</details>
	* <details>
		<summary>gob encoding vs protobuf with preallocating modifications</summary>
		
		```
		goos: darwin
		goarch: arm64
		pkg: github.com/grafana/mimir/pkg/storegateway
		                                                                                                                           │  before.txt   │ preallocate-label-values-slice-28024ebdb.txt │
		                                                                                                                           │    sec/op     │       sec/op         vs base                 │
		BucketStoreLabelValues/inmemory_cache/10-series-matched-with-10-label-values-10                                               73.76µ ± ∞ ¹          18.11µ ± ∞ ¹  -75.44% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1000-series-matched-with-1000-label-values-10                                          119.91µ ± ∞ ¹          90.71µ ± ∞ ¹  -24.35% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-10-label-values-10                                        73.94µ ± ∞ ¹          18.26µ ± ∞ ¹  -75.30% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1-label-values-10                                         72.29µ ± ∞ ¹          18.12µ ± ∞ ¹  -74.94% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1_000_000-label-values-10                                                       34.87m ± ∞ ¹
		geomean                                                                                                                       1.156m                113.6µ        -67.25%               ²
		¹ need >= 6 samples for confidence interval at level 0.95
		² benchmark set differs from baseline; geomeans may not be comparable
		
		                                                                                                                           │   before.txt   │ preallocate-label-values-slice-28024ebdb.txt │
		                                                                                                                           │      B/op      │        B/op          vs base                 │
		BucketStoreLabelValues/inmemory_cache/10-series-matched-with-10-label-values-10                                              51.391Ki ± ∞ ¹         7.430Ki ± ∞ ¹  -85.54% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1000-series-matched-with-1000-label-values-10                                           213.7Ki ± ∞ ¹         153.5Ki ± ∞ ¹  -28.17% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-10-label-values-10                                       50.951Ki ± ∞ ¹         7.017Ki ± ∞ ¹  -86.23% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1-label-values-10                                        51.043Ki ± ∞ ¹         6.969Ki ± ∞ ¹  -86.35% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1_000_000-label-values-10                                                       125.0Mi ± ∞ ¹
		geomean                                                                                                                       1.464Mi               93.48Ki        -78.98%               ²
		¹ need >= 6 samples for confidence interval at level 0.95
		² benchmark set differs from baseline; geomeans may not be comparable
		
		                                                                                                                           │  before.txt  │ preallocate-label-values-slice-28024ebdb.txt │
		                                                                                                                           │  allocs/op   │      allocs/op       vs base                 │
		BucketStoreLabelValues/inmemory_cache/10-series-matched-with-10-label-values-10                                              1343.0 ± ∞ ¹           173.0 ± ∞ ¹  -87.12% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1000-series-matched-with-1000-label-values-10                                          4.314k ± ∞ ¹          3.144k ± ∞ ¹  -27.12% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-10-label-values-10                                       1342.0 ± ∞ ¹           172.0 ± ∞ ¹  -87.18% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1-label-values-10                                        1330.0 ± ∞ ¹           166.0 ± ∞ ¹  -87.52% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1_000_000-label-values-10                                                      3.000M ± ∞ ¹
		geomean                                                                                                                      5.945k                2.156k        -80.31%               ²
		¹ need >= 6 samples for confidence interval at level 0.95
		² benchmark set differs from baseline; geomeans may not be comparable
		```
		
		</details>

* custom binary encoding
	* after all we're just writing 2 strings and then a slice of strings; this is simple enough to implement and shows the best perf
	* <details>
		<summary>protobuf vs custom encoding</summary>
		
		```
		goos: darwin
		goarch: arm64
		pkg: github.com/grafana/mimir/pkg/storegateway
		                                                                                              │ preallocate-label-values-slice-28024ebdb.txt │    custom-encoding-a8267773c.txt    │
		                                                                                              │                    sec/op                    │    sec/op     vs base               │
		BucketStoreLabelValues/inmemory_cache/10-series-matched-with-10-label-values-10                                                 18.11µ ± ∞ ¹   17.62µ ± ∞ ¹   -2.76% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1000-series-matched-with-1000-label-values-10                                             90.71µ ± ∞ ¹   65.81µ ± ∞ ¹  -27.45% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-10-label-values-10                                          18.26µ ± ∞ ¹   17.32µ ± ∞ ¹   -5.15% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1-label-values-10                                           18.12µ ± ∞ ¹   17.02µ ± ∞ ¹   -6.09% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1_000_000-label-values-10                                   34.87m ± ∞ ¹   23.54m ± ∞ ¹  -32.49% (p=0.029 n=4)
		geomean                                                                                                                         113.6µ         95.74µ        -15.76%
		¹ need >= 6 samples for confidence interval at level 0.95
		
		                                                                                              │ preallocate-label-values-slice-28024ebdb.txt │    custom-encoding-a8267773c.txt     │
		                                                                                              │                     B/op                     │     B/op       vs base               │
		BucketStoreLabelValues/inmemory_cache/10-series-matched-with-10-label-values-10                                                7.430Ki ± ∞ ¹   7.145Ki ± ∞ ¹   -3.83% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1000-series-matched-with-1000-label-values-10                                            153.5Ki ± ∞ ¹   143.5Ki ± ∞ ¹   -6.56% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-10-label-values-10                                         7.017Ki ± ∞ ¹   6.826Ki ± ∞ ¹   -2.72% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1-label-values-10                                          6.969Ki ± ∞ ¹   6.453Ki ± ∞ ¹   -7.40% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1_000_000-label-values-10                                 124.98Mi ± ∞ ¹   99.43Mi ± ∞ ¹  -20.44% (p=0.029 n=4)
		geomean                                                                                                                        93.48Ki         85.61Ki         -8.42%
		¹ need >= 6 samples for confidence interval at level 0.95
		
		                                                                                              │ preallocate-label-values-slice-28024ebdb.txt │    custom-encoding-a8267773c.txt    │
		                                                                                              │                  allocs/op                   │  allocs/op   vs base                │
		BucketStoreLabelValues/inmemory_cache/10-series-matched-with-10-label-values-10                                                  173.0 ± ∞ ¹   161.0 ± ∞ ¹    -6.94% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1000-series-matched-with-1000-label-values-10                                             3144.0 ± ∞ ¹   162.0 ± ∞ ¹   -94.85% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-10-label-values-10                                           172.0 ± ∞ ¹   160.0 ± ∞ ¹    -6.98% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1-label-values-10                                            166.0 ± ∞ ¹   154.0 ± ∞ ¹    -7.23% (p=0.029 n=4)
		BucketStoreLabelValues/inmemory_cache/1_000_000-series-matched-with-1_000_000-label-values-10                                3000116.0 ± ∞ ¹   140.0 ± ∞ ¹  -100.00% (p=0.029 n=4)
		geomean                                                                                                                         2.156k         155.2         -92.80%
		¹ need >= 6 samples for confidence interval at level 0.95
		```
		</details>


My preference is for avoiding caching these large responses in the first place. It's doesn't involve any bug-prone code and is shorter to implement.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -38,6 +38,7 @@
 * [BUGFIX] Ruler: Support the `type=alert|record` query parameter for the API endpoint `<prometheus-http-prefix>/api/v1/rules`. #4302
 * [BUGFIX] Backend: Check that alertmanager's data-dir doesn't overlap with bucket-sync dir. #4921
 * [BUGFIX] Alertmanager: Allow to rate-limit webex, telegram and discord notifications. #4979
+* [BUGFIX] Store-gateway: panics when decoding LabelValues responses that contain more than 655360 values. These responses are no longer cached. #5021
 
 ### Documentation
 

--- a/pkg/storegateway/bucket.go
+++ b/pkg/storegateway/bucket.go
@@ -1384,6 +1384,12 @@ func fetchCachedLabelValues(ctx context.Context, indexCache indexcache.IndexCach
 }
 
 func storeCachedLabelValues(ctx context.Context, indexCache indexcache.IndexCache, userID string, blockID ulid.ULID, labelName string, matchers []*labels.Matcher, values []string, logger log.Logger) {
+	// This limit is a workaround for panics in decoding large responses. See https://github.com/golang/go/issues/59172
+	const valuesLimit = 655360
+	if len(values) > valuesLimit {
+		level.Debug(spanlogger.FromContext(ctx, logger)).Log("msg", "skipping storing label values response to cache because it exceeds number of values limit", "limit", valuesLimit, "values_count", len(values))
+		return
+	}
 	entry := labelValuesCacheEntry{
 		Values:      values,
 		LabelName:   labelName,

